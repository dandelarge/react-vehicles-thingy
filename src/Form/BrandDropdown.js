import React from 'react'
import Dropdown from '../designSystem/Dropdown';

export default function BrandDropdown({optionsList, value, onChange}) {
    const typesDropdown = {
        fieldName: 'brand',
        fieldId: 'brand',
        label: 'Brand',
        optionsList: [{label: 'pick a brand', value: ''}, ...optionsList],
        value,
        onChange
    };

    return (
        <Dropdown {...typesDropdown} />
    )
}
