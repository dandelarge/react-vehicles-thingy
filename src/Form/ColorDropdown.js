import React from 'react'
import Dropdown from '../designSystem/Dropdown';

export default function ColorDropdown({optionsList, value, onChange}) {
    const typesDropdown = {
        fieldName: 'color',
        fieldId: 'color',
        label: 'Color',
        optionsList: [{label: 'pick a color', value: '', disabled: true}, ...optionsList],
        value,
        onChange
    };

    return (
        <Dropdown {...typesDropdown} />
    )
}
