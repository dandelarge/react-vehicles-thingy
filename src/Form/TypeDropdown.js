import React from 'react'
import Dropdown from '../designSystem/Dropdown';

export default function TypeDropdown({optionsList, value, onChange}) {
    const typesDropdown = {
        fieldName: 'type',
        fieldId: 'type',
        label: 'Vehicle Type',
        optionsList: [{label: 'pick a type of vehicle', value: '', disabled: true}, ...optionsList],
        value,
        onChange
    };

    return (
        <Dropdown {...typesDropdown} />
    )
}
