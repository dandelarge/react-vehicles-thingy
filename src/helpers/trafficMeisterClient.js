import trafficMeister from '../service';

export function fetchData() {
    return new Promise((resolve, reject) => {
        trafficMeister.fetchData((err, data) => {
            if(err) {
                reject(new Error(err));
            }
            resolve(data);
        });
    });
}
