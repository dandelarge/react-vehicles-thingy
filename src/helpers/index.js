import { fetchData } from './trafficMeisterClient';

export function makeTypesList(list) {
    let types = {};
    const typesArr = list
        .map(item => ({label: item.type, value: item.type}))
        .filter((val, index) => {
            if(val.value in types) {
                return false;
            }
            types[val.value] = index;
            return true;
        });
    return typesArr;
};

export function makeBrandOptions(list, type) {
    return list
        .filter(vehicle => vehicle.type === type)
        .map(vehicle => ({label: vehicle.brand, value: vehicle.brand}));
};

export const trafficMeisterClient = {
    fetchData
};