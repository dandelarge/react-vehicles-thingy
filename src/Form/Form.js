import React, {useState, useEffect} from 'react'

import {makeTypesList, makeBrandOptions, trafficMeisterClient} from '../helpers';
import TypeDropdown from './TypeDropdown';
import BrandDropdown from './BrandDropdown';
import ColorDropdown from './ColorDropdown';
import Spinner from '../designSystem/Spinner';

export default function Form(props) {

    // Setting loading state
    const [loading, setLoading] = useState(false);

    //Model
    const [vehiclesList, setVehiclesList] = useState([]);

    //Select values
    const [selectedType, setSelectedType] = useState('');
    const [selectedBrand, setSelectedBrand] = useState('');
    const [selectedColor, setSelectedColor] = useState('');

    //Options for the selects
    const [typesOptions, setTypesOptions] = useState([]);
    const [brandOptions, setBrandOptions] = useState([]);
    const [colorOptions, setColorOptions] = useState([]);

    // Handling initial rendering
    useEffect(() => {
        let errorCount = 0;
        (async function onMount(retries = 3) {
            setLoading(true);
            try {
                const list = await trafficMeisterClient.fetchData();
                setVehiclesList(list);
                setTypesOptions(makeTypesList(list));
                setLoading(false);
            } catch(e) {
                // If the API fails, we try again a few more times
                if (errorCount < retries) {
                    errorCount++;
                    onMount(retries);
                // if it doesn't reply after few tries, we stop the spinner
                } else {
                    setLoading(false);
                    console.error(e);
                }
            }
        })();
    }, []);

    useEffect(() => {
        setSelectedBrand('');
        setSelectedColor('');
    }, [selectedType, vehiclesList]);

    // Callbacks to Handle Change
    function onTypeSelect(e)  {
        setSelectedType(e.target.value);
        setBrandOptions(makeBrandOptions(vehiclesList, e.target.value));
        props.setData(e.target.value);
        props.getImage('/img/car-icon.jpg');
        props.setError(false);
    }

    function onBrandSelect(e) {
        const obj = vehiclesList.filter(vehicle => vehicle.brand === e.target.value)[0];
        setSelectedBrand(e.target.value);
        setColorOptions(obj.colors.map(color => ({label: color, value: color})));
        props.setData(selectedType, e.target.value);
        props.getImage(obj.img);
        props.setError(false);
    }

    function onColorSelect(e) {
        setSelectedColor(e.target.value);
        props.setData(selectedType, selectedBrand, e.target.value);
    }

    return (
        <form>
          <h3>{props.title}</h3>
          <TypeDropdown optionsList={typesOptions} value={selectedType} onChange={onTypeSelect}/>
          <BrandDropdown optionsList={brandOptions} value={selectedBrand} onChange={onBrandSelect}/>
          <ColorDropdown optionsList={colorOptions} value={selectedColor} onChange={onColorSelect} />
          {!loading || <Spinner />}
        </form>
    )
}
