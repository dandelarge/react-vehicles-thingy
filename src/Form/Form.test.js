import React from 'react';
import { render, getByTestId, act, fireEvent } from 'react-testing-library';
import Form from './Form';
import { trafficMeisterClient } from '../helpers';

describe('Form', () => {
    const clientSpy = jest.spyOn(trafficMeisterClient, 'fetchData');
    const mockedList = [{
        type: 'car',
        brand: 'Bugatti Veyron',
        colors: ['red', 'black'],
        img: 'https://upload.wikimedia.org/wikipedia/commons/thumb/c/c9/Bugatti_Veyron_16.4_%E2%80%93_Frontansicht_%281%29%2C_5._April_2012%2C_D%C3%BCsseldorf.jpg/520px-Bugatti_Veyron_16.4_%E2%80%93_Frontansicht_%281%29%2C_5._April_2012%2C_D%C3%BCsseldorf.jpg'
    }];

    let formWrapper;

    beforeEach(() => {
        act(() => {
            clientSpy.mockResolvedValue(mockedList);
            formWrapper = render(<Form
                title="This is the test!"
                getImage={img => console.log(img)}
                setData={data => console.log(data)}
                setError={something => console.log(something)} />);
        });

    });

    it('should actually render', () => {
        expect(formWrapper.container).toMatchSnapshot();
    })

    it('should try to retrieve data on load', () => {
        expect(clientSpy).toHaveBeenCalled();
    });

    it('should get one type of vehicle', () => {
        const typeSelect = getByTestId(formWrapper.container, 'type-select');
        expect(typeSelect.childNodes.length).toEqual(2);
    });

    it('should call the API again in case the first call fails', () => {
        clientSpy.mockRejectedValueOnce({
            error: "some weird error"
        });

        render(<Form title="This is the test!" />);

        //5 because it's called once per each test, and twice here.
        expect(clientSpy).toHaveBeenCalledTimes(5);
    });

});