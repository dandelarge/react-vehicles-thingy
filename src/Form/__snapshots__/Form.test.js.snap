// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should actually render 1`] = `
<div>
  <form>
    <h3>
      This is the test!
    </h3>
    <div
      class="fieldset"
    >
      <select
        class="App-dropdown"
        data-testid="type-select"
        id="type"
        name="type"
      >
        <option
          disabled=""
          value=""
        >
          pick a type of vehicle
        </option>
        <option
          value="car"
        >
          car
        </option>
      </select>
      <label
        for=""
      >
        Vehicle Type
      </label>
    </div>
    <div
      class="fieldset"
    >
      <select
        class="App-dropdown"
        data-testid="brand-select"
        id="brand"
        name="brand"
      >
        <option
          disabled=""
          value=""
        >
          pick a brand
        </option>
      </select>
      <label
        for=""
      >
        Brand
      </label>
    </div>
    <div
      class="fieldset"
    >
      <select
        class="App-dropdown"
        data-testid="color-select"
        id="color"
        name="color"
      >
        <option
          disabled=""
          value=""
        >
          pick a color
        </option>
      </select>
      <label
        for=""
      >
        Color
      </label>
    </div>
  </form>
</div>
`;
